pkg/dump/dumper.go

```diff
@@ -0,0 +1,430 @@
+	}
+
+	if err := n.Close(); err != nil {
+		log.Printf("error closing connection: %v", err)
+	}
+
+	return nil
+}
+
+// sshClient is an interface abstracting *ssh.Client, which allows us to test it
+type sshClient interface {
+	io.Closer
+
+	// ExecPiped runs the command, piping stdout & stderr
+	ExecPiped(ctx context.Context, command string, stdout io.Writer, stderr io.Writer) error
+}
+
+// sshClientFactory is an interface abstracting to a node over SSH
+type sshClientFactory interface {
+	Dial(ctx context.Context, host string) (sshClient, error)
+}
+
+// logDumperNode holds state for a particular node we are dumping
+type logDumperNode struct {
+	client sshClient
+	dumper *logDumper
+
+	dir string
+}
+
+// connectToNode makes an SSH connection to the node and returns a logDumperNode
+func (d *logDumper) connectToNode(ctx context.Context, nodeName string, host string) (*logDumperNode, error) {
+	client, err := d.sshClientFactory.Dial(ctx, host)
+	if err != nil {
+		return nil, fmt.Errorf("unable to SSH to %q: %v", host, err)
+	}
+	return &logDumperNode{
+		client: client,
+		dir:    filepath.Join(d.artifactsDir, nodeName),
+		dumper: d,
+	}, nil
+}
+
+// logDumperNode cleans up any state in the logDumperNode
+func (n *logDumperNode) Close() error {
+	return n.client.Close()
+}
+
+// dump captures the well-known set of logs
+func (n *logDumperNode) dump(ctx context.Context) []error {
+	if ctx.Err() != nil {
+		return []error{ctx.Err()}
+	}
+
+	var errors []error
+
+	// Capture kernel log
+	if err := n.shellToFile(ctx, "sudo journalctl --output=short-precise -k", filepath.Join(n.dir, "kern.log")); err != nil {
+		errors = append(errors, err)
+	}
+
+	// Capture full journal - needed so we can see e.g. disk mounts
+	// This does duplicate the other files, but ensures we have all output
+	if err := n.shellToFile(ctx, "sudo journalctl --output=short-precise", filepath.Join(n.dir, "journal.log")); err != nil {
+		errors = append(errors, err)
+	}
+
+	// Capture logs from any systemd services in our list, that are registered
+	services, err := n.listSystemdUnits(ctx)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("error listing systemd services: %v", err))
+	}
+	for _, s := range n.dumper.services {
+		name := s + ".service"
+		for _, service := range services {
+			if service == name {
+				if err := n.shellToFile(ctx, "sudo journalctl --output=cat -u "+name, filepath.Join(n.dir, s+".log")); err != nil {
+					errors = append(errors, err)
+				}
+			}
+		}
+	}
+
+	// Capture any file logs where the files exist
+	fileList, err := n.findFiles(ctx, "/var/log")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("error reading /var/log: %v", err))
+	}
+	for _, name := range n.dumper.files {
+		prefix := "/var/log/" + name + ".log"
+		for _, f := range fileList {
+			if !strings.HasPrefix(f, prefix) {
+				continue
+			}
+			if err := n.shellToFile(ctx, "sudo cat "+f, filepath.Join(n.dir, filepath.Base(f))); err != nil {
```

Comment:
If `f` happens to have shell metacharacters then they will be interpreted as such. I believe `f` needs to be quoted for use in shell.